pkg/controllers/node/controller.go

```diff
@@ -76,51 +70,48 @@ func NewNodeController(ctx context.Context, k8sClientset *kubernetes.Clientset,
 		ctx:          ctx,
 		calicoClient: calicoClient,
 		k8sClientset: k8sClientset,
-		rl:           workqueue.DefaultControllerRateLimiter(),
-		nodemapper:   map[string]string{},
-		allBlocks:    map[string]model.KVPair{},
-		config:       cfg,
-		nodeCache:    make(map[string]*api.Node),
+		dataFeed:     NewDataFeed(calicoClient),
 	}
 
-	// channel used to kick the controller into scheduling a sync. It has length
-	// 1 so that we coalesce multiple kicks while a sync is happening down to
-	// just one additional sync.
-	nc.schedule = make(chan interface{}, 1)
-	nc.blockUpdate = make(chan interface{}, 1)
-
 	// Create a Node watcher.
 	listWatcher := cache.NewListWatchFromClient(k8sClientset.CoreV1().RESTClient(), "nodes", "", fields.Everything())
 
+	// Create the IPAM controller.
+	nc.ipamCtrl = NewIPAMController(cfg, calicoClient, k8sClientset)
+	nc.ipamCtrl.RegisterWith(nc.dataFeed)
+
 	// Setup event handlers
 	handlers := cache.ResourceEventHandlerFuncs{
 		DeleteFunc: func(obj interface{}) {
-			// Just kick controller to wake up and perform a sync. No need to bother what node it was
-			// as we sync everything.
-			kick(nc.schedule)
+			nc.ipamCtrl.OnKubernetesNodeDeleted()
 		}}
 
-	// Note that the configuration code has already handled disabling this if
-	// we are in KDD mode.
+	// Create the Auto HostEndpoint sub-controller and register it to receive data.
+	// We always launch this controller, even if auto-HEPs are disabled, since the controller
+	// is responsible for cleaning up after itself in case it was previously enabled.
+	autoHEPController := NewAuthoHEPController(cfg, calicoClient)
```

Comment:
Is this a typo? `NewAuthoHEPController` -> `NewAutoHEPController`

Reply:
yeeeeep